### ISSUE TITLE
ISBLOCKHASH

### ISSUE THREAD

**Issue by fulldecent** [ethereum/EIPs#3543]
Blockhashes are an important source of randomness in Ethereum and they are secure if used in a commit-reveal scheme that includes enough blocks.

Secure coin flipping implementation:

1. COMMIT to accept any outcome that is available 5 blocks from now.
2. REVEAL result of BLOCKHASH[NUMBER] ^ BLOCKHASH[NUMBER-1]  ^ BLOCKHASH[NUMBER-2] ...

The cost of attacking this is the cost of making five consecutive blocks.

Five can be arbitrarily large for arbitrary security.

This can be particularly useful not just for simple smart contracts, but also for rollups, chains and other layer 2 solutions.

---

The BLOCKHASH function only provides the recent 256 hashes. This limits applications to being useful only under a COMMIT-REVEAL timespan of 60 ~minutes.

And extending BLOCKHASH to return any hash would cost O(N) storage on light validating nodes, unacceptable.

This proposal allows to get blockhashes using O(LOG(N)) node storage and O(LOG(N)) computation, which is billed appropriately.

---

NEW OP CODE: ISBLOCKHASH

1. Argument 1: the number of the block to consider
2. Argument 2: a merkle tree proof from `blocknum`'s blockhash up through the root at 2^CEIL(LOG2(`blocknum`))

The result of the opcode is 1 if the merkle tree proof of length `i` results in the root at 2^CEIL(LOG2(`blocknum`)). Zero otherwise. Also zero if the proof alleges that the block hash is zero. (i.e. it is in the future)

Implementation:

Clients shall encode a root hash of merkle tree of each blockhashes in each block. (Future blocks have a "hash" of zero in this tree.) This costs LOG(N) storage to recalculate an append-only binary structure that is summarized in a merkle root.

Gas cost:

Cost is proportional to the proof length plus a constant.

**Comment by zemse** [comment]
If an attacker wants to influence the coin flip, instead of spending resources for all 5 blocks, they can only put all of their resources to mine the last block with a block hash that is beneficial for them. Since while mining the last block, the previous 4 block hashes are already mined, public and likely fixed (may change if reorg happens). In that case, `REVEAL` should be just as random as the last block hash.

**Comment by fulldecent** [comment]
Yup! Rookie mistake there. Best you can ever get is one block of entropy for one decision.

The benefit is only to have multiple random things pending.

**Comment by Shymaa-Arafat** [comment]
Sorry if this out of scope of ur issue, but it could be related since I was thinking is the Ethereum susceptible to a similar hash collision attack as the onedescribed in the Utreexo 2019 original paper ...
If the attacker in a position to know all the Trie leaves then the complexity of faking one reduces by the number of leaves. For example for a 256bit hash and say 2³² leaves it reduces  from 2²⁵⁶ to 2²²⁴. 
![IMG_20210430_203147](https://user-images.githubusercontent.com/83260375/116742864-4afcd800-a9f8-11eb-8fb6-80dce950ab46.jpg)
He solves this by concatenating the block hash (and suggests position but that's another story)
![IMG_20210430_203553](https://user-images.githubusercontent.com/83260375/116742947-797ab300-a9f8-11eb-9453-2a4d9370955e.jpg)
.
Now I was thinking does concatenating the key hash in EIP-3102 have the same effect?
https://eips.ethereum.org/EIPS/eip-3102
Where it is stated that
leaf_hash = hash(hash(key) || hash(leaf_value))
Is "key" here an address or what?and how many bits?
That's what is written there...
![IMG_20210429_202939](https://user-images.githubusercontent.com/83260375/116743229-eb52fc80-a9f8-11eb-8854-c4fdcf4e8764.jpg)
.
Hope my thought is in the scope of ur issue.

**Comment by fulldecent** [comment]
I believe there are other world matters that eclipse this concern. 2,000 years is a long time (how long 2^32 blocks on Ethereum takes).

**Comment by Shymaa-Arafat** [comment]
My example was about 2³² nodes in the Trie, not blocks?!!!!!!!!!!!!

**Comment by fulldecent** [comment]
Right. In this EIP, each block is a node in the Trie.

**Comment by Shymaa-Arafat** [comment]
I'm sorry which EIP do u mean? 
EIP-3102 or what?
Aren't there always a storage Trie and a state Trie?
![IMG_20210419_004925](https://user-images.githubusercontent.com/83260375/116840412-fdd65d00-abd5-11eb-84a3-228d03942e8f.jpg)
(From
https://blog.ethereum.org/2020/11/30/the-1x-files-code-merkleization/)
